pandas/tseries/index.py

```diff
@@ -1257,6 +1257,12 @@ def get_loc(self, key, method=None):
                                           'when key is a time object')
             return self.indexer_at_time(key)
 
+        # check if it's a Period and the frequencies are the same - otherwise a monthly period would match for
+        # a daily timestamp at the beginning of the month. NB: 'B' and 'D' therefore won't match
+        if isinstance(key, com.ABCPeriod) and key.freq == self.freq:
```

Comment:
@jreback Do you think any of these should be `True`? I had thought that they should all be `False` - that unless there in an explicit periodicity / freq to a `DatetimeIndex`, it couldn't match a `Period`. 
I'm pushing that in the last case, but it seems that if you start guessing there you enter a murky gray area, and a better way of operating is for the user to specify the `freq` for explicitness 

```
In [7]: pd.date_range('2013-01',periods=10,freq='M')[2:4:2]
Out[7]: DatetimeIndex(['2013-03-31'], dtype='datetime64[ns]', freq='2M', tz=None)

In [8]: pd.Period('2013-03-31','D') in _
Out[8]: False
```

```
In [9]: pd.date_range('2013-01', periods=10, freq='D')[::2]
Out[9]: 
DatetimeIndex(['2013-01-01', '2013-01-03', '2013-01-05', '2013-01-07',
               '2013-01-09'],
              dtype='datetime64[ns]', freq='2D', tz=None)

In [10]: pd.Period('2013-01-01','D') in _
Out[10]: False
```
```
In [11]: pd.date_range('2013-01',periods=10,freq='M')
Out[11]: 
DatetimeIndex(['2013-01-31', '2013-02-28', '2013-03-31', '2013-04-30',
               '2013-05-31', '2013-06-30', '2013-07-31', '2013-08-31',
               '2013-09-30', '2013-10-31'],
              dtype='datetime64[ns]', freq='M', tz=None)

In [12]: pd.Period('2013-10-31','D') in _
Out[12]: False
```
```
In [17]: idx=pd.date_range('2013-01',periods=10)

In [18]: idx.freq=None

In [19]: idx
Out[19]: 
DatetimeIndex(['2013-01-01', '2013-01-02', '2013-01-03', '2013-01-04',
               '2013-01-05', '2013-01-06', '2013-01-07', '2013-01-08',
               '2013-01-09', '2013-01-10'],
              dtype='datetime64[ns]', freq=None, tz=None)

In [20]: pd.Period('2013-01-01', 'B') in idx
Out[20]: False

In [21]: pd.Period('2013-01-01', 'D') in idx
Out[21]: False
```